data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -112,8 +112,15 @@
   /**
    * Supplier for cycle checker used when traversing instances using a {@link Data.TraverseCallback}. Applications can
    * choose to replace this with a supplier vending custom implementations, at their own risk of ensuring correctness.
+   *
+   * <p>The default implementation uses a {@link ThreadLocal} in every {@link DataList} and {@link DataMap} to detect
+   * and record cycles, when assertions are enabled, and does NO checks when assertions are disabled.</p>
```

Comment:
nit: record cycles when assertions are .. (remove comma)

Reply:
Also made the checkers singletons to reduce garbage since they carry no internal state.